### PR TITLE
the browser's K-quant dot rides the idot lattice, wasm gets its FMA, and neither the spin window nor the audio mixer pays the host clock

### PR DESCRIPTION
**Pull request by borisbat**
**Behavior changes: the browser's K-quant GEMV is SIMD, a wasm build emits FMA, a browser's job-queue workers spin, and the browser audio mixer no longer corrupts on Windows.**

**Why.** The browser tier ran the portable kernels, and the portable K-quant dot walked its weights one at a time while the `idot` lattice op already had a lowering for every target. Separately, a wasm build had no multiply-add at all, and the job queue's spin window and dasAudio's mixer both paid a JS crossing per iteration for the clock - which on Windows is a 15.6 ms quantum, small enough to empty the audio ring every cycle.

**What changes.**

- `dot_k4q8` splits each packed byte into two nibbles on a SIGNED `byte16` and feeds `idot` - one body, no target branch.
- `+relaxed-simd` joins the wasm feature string, so a contracted multiply-add becomes `f32x4.relaxed_madd` instead of splitting.
- `idot_wasm_simd128` takes the operand signedness and widens an unsigned operand with `ZExt`, keeping the `us` form off the 16-lane generic path.
- The spin window reads the deadline clock once per stride and pauses without entering JS; a zero window still parks on the first test.
- A browser's workers take the one 30 ms spin window every other target takes: the emscripten `[init]` that zeroed it is deleted, so `setup_dasllama_jobque` pushes the default with no platform test.
- dasAudio's browser mixer parks on a drain counter instead of a sleep, primes its ring before starting the device, and counts underruns.
- Two drive-bys, because this branch's own gate run is what found them: the `standalone_ctx` shared fixture lands where its generated bindings look for it, so `standalone_capi_dasbind` can pass under a multi-config generator; and preflight's compile sweep stops trying to compile `examples/dasLLAMA/library/main.das`, whose `dasllama_lib_c` an `EXCLUDE_FROM_ALL` target writes.

**Observable behavior.**

- Pocket TTS on the parrot page reads 1.1x -> 2.3x real time under one protocol: 1.7x from the kernel, then 2.3x with the workers spinning. In the author's own Chrome the same build reads 4.7x.
- `dot_k4q8` is 1.90x on x64, and bit-exact against the scalar body on both targets.
- Browser audio no longer clicks on Windows, and `sound_playback_underrun_frames` reports when a mixer does miss the device.
- Output latency is 20 ms where it was 10.7 ms; the parrot lab opens with the spin toggle on.
- A browser session that preferred parking must now ask for it - `set_jobque_spin_us(0)` before `setup_dasllama_jobque`, or `jobque_spin_us` in a profile.

**Where to look.** The kernel is `modules/dasLLAMA/dasllama/dasllama_math_default.das`; the spin default is `modules/dasLLAMA/dasllama/dasllama_math.das`; the wasm codegen is `modules/dasLLVM/daslib/`; the spin window is `src/misc/job_que.cpp`; the mixer is `modules/dasAudio/src/dasAudio.cpp`.

<details>
<summary>Validation, claims, ledger</summary>

### Validation

- dasLLAMA `--suite model-free`, re-run AFTER the spin-default flip (the earlier run predated it): 74 files, 0 failed, 5 cells skipped for an absent device or model, each loudly. `test_prefill_cpu_kernels` 30/30, `test_tts_blocks` 29/29, `test_kqformat` 22/22.
- `test_kquant.das`, run directly because it is the `dot_k4q8` gate and it sits in the `stocked` suite: 287 tests, 282 passed, 5 skipped for absent models, 0 failed - including `test_kq_dots`, which holds every `dot_*q8` body against an fp64 plane-dequant reference at a 1e-4 relative bar, at n=1024 and n=3072 as well as the base widths.
- `test_facade.das`, run directly for the same reason: 9 tests, 6 passed, 3 skipped for absent models, 0 failed. Its `test_facade_jobque_spin_knob` cell needs no model, so the flipped 30 ms default is proven rather than assumed.
- `tests/jobque`: 263 tests, 263 passed - the suite that covers the spin-window change.
- `preflight --full`, then the lane gates re-run targeted: `docs` (das2rst, imgui2rst, vulkan2rst, stubs, uncategorized, untracked, sphinx-html) PASS, `tests-aot` PASS, `ast-verify`, `ci-das`, `ci-matrix`, `dasgen`, `hash-refs`, `review-md`, `review-md-tests`, `md-ascii`, `format`, `untracked` PASS. `tests-interp` 14523 tests / 14507 passed, `tests-jit` 14391 / 14376, `tests-cpp` 153 ctest cases - every remaining red on this box accounted for below.
- `modules/dasLLVM/tests` (the folder CI's `--test tests` does not reach, and which this diff's emitter change owes): 116 tests, 115 passed. The one failure, `llvm_vector_math.das::test_vector_log_special_values`, is PRE-EXISTING on master - restoring master's four emitter files into the tree reproduces it exactly, and CI's `extended_checks` runs that same file on darwin and passes, so it is an x64-host lowering divergence, not this branch.
- `tests-cpp`'s one failure was a real bug and is FIXED here: `standalone_capi_dasbind` could never pass under a multi-config generator, because `daslang -ctx -bindings_library "${CMAKE_CURRENT_BINARY_DIR}/standalone_init_fixture"` bakes a config-less path into the generated bindings while `RUNTIME_OUTPUT_DIRECTORY` gains a `Release/` subdirectory there. Both output directories now use `$<1:...>`, which suppresses the per-config subdirectory; the case re-runs 1/1 Passed.
- Two reds on this box are environment, not code, and both pass when the environment is right: `jit_lib.das::test_jit_lib_static_and_export_all` needs MSVC's `lib.exe` on PATH (7 tests / 7 passed under `vcvars64`), and `tests/watchdog/test_watchdog.das` passes 59/59 but takes 74.6 s against a 60 s per-file cap that is preflight-local, not a CI gate.
- `utils/benchctl/tests/` is red on Windows and green on CI's Linux; it arrived with `43b942576`, two commits before this branch, and nothing here touches it. `test_bench_cli.das:23` spawns `"bin/daslang"` where the same file passes `get_host_binary()` three lines below, and `test_bench_runner.das`'s unrunnable-binary case dies because `spawn_process` THROWS on a failed `CreateProcess` while POSIX `fork` succeeds and reports through the child's exit code - which also makes `bench_runner.das`'s `process_pid(p) <= 0` guard dead code on both platforms. Left for a ruling on that builtin's contract rather than patched blind from here.
- Lint over the changed set is clean on both rails, including LINT026/LINT027 over `modules/dasLLVM`, `modules/dasAudio` and `src/misc`. A pre-existing STYLE025 in `llvm_jit_common.das` and another in `tests/jit_tests/wasm_idot_lowering.das`, both of which the changed-files gate surfaces, are fixed.
- The artifact behind every browser figure is a `-exe --jit-target=wasm64-unknown-emscripten` build - `daspkg release wasm` over `examples/dasLLAMA/parrot`, served locally to Chrome. That is what exercises the target-triple branches this diff adds: `wasm_target_features`'s `+relaxed-simd`, `idot_wasm_simd128`'s ZExt arm, and `job_que.cpp`'s emscripten stride and pause. The host suite never reaches them; `tests/jit_tests/wasm_idot_lowering.das` reaches the IR for the first two by cross-compiling, not by running.
- Browser figures are the engine's own per-synthesis timing line read from the page console, with the window verified in the foreground before AND after every run and both builds measured back to back in one session differing only in the change under test.
- x64 figures are both kernel bodies compiled into one process and timed round-robin so frequency drift cancels, best of nine, with the browser closed.
- The Windows timer quantum is measured, not assumed: a 1 ms `Atomics.wait` in a Worker on this box returns in 16.0 ms median, 16.6 ms worst, against about 1 ms on macOS.
- Audio verified by ear on storywisher, storyteller and parrot at 20 ms, with the underrun counter reading zero.
- dasLLAMA `--suite stocked` was NOT run as a suite, by the author's ruling: the kernel this diff rewrites is the wasm/reference body, so nothing the stocked models exercise differs from master. Two of its files are exceptions that ruling did not anticipate, and both are run above: `test_kquant.das` IS the `dot_k4q8` gate, and `test_facade.das` carries the spin-default assertion this change edits.
- `compile-sweep`'s one red is the second drive-by above. `examples/dasLLAMA/library/main.das` requires `dasllama_lib_c`, written only by the `EXCLUDE_FROM_ALL` `dasllama_lib_jit` target - "Opt-in … ten minutes and ~6 GB to compile", per that CMakeLists' own header - so the root is unresolvable on any box that has not paid that build. It is the `SWEEP_EXCLUDED` charter's own case ("a resolution the plain compile cannot do") and was simply missing its row; the row has a test cell plus the negative control that `parrot/main.das` is still swept.

### Claims - stated, not tested

- Bit-exactness is not what the suite gates. The K-quant kernel tests compare against a dequant reference at 1e-4, so a body that merely rounded differently would still pass. The claim rests on the direct comparison instead: both bodies compiled into one process over the same weights produce the identical float sum, -77683816.
- The browser mixer path compiles only under emscripten with pthreads, so nothing native executes it and no CI lane reaches it. Its correctness here rests on the measured quantum, the browser runs above, and the two `REVIEW.md` rules - not on a test. Making that path natively testable is the follow-up this change does not take.
- CI runs Linux, whose timer quantum matches macOS, so a page-level test there would catch a lost wake or a dead producer but not the class of defect fixed here.

### Not done

- No board row. Every browser figure is direction-grade - an rtf read from a page console, not an `lcpp_bench` cell - so `performance/records/zen2.json` gains nothing from this arc, and `PERF_LEDGER.md` says so inside the entry.
- The k4/k5 fold. `dot_k4q8` and `dot_k5q8` share their superblock walk, the Q4_K scale-block reads and both integer folds, and differ only in how each reconstructs a quant - and this change widens that gap, since only k4 now rides `idot`. The dupe audit proposed parameterizing the shared scaffolding. Declined here: the fold rewrites `dot_k5q8`, which this arc never measured, and the scope is the kernel the example runs.
- A native `sound_playback_underrun_frames`. It answers a real count under emscripten with pthreads and a hard 0 everywhere else, so a script cannot tell "no underruns" from "this target does not count them". `PERF_LEDGER.md` carries it as OWED.
- The relaxed-SIMD dot is still not what the generic `idot` lowers to - its second operand is 7-bit, which the lattice op cannot promise.

</details>

### CODE REVIEW

File: modules/dasAudio/src/dasAudio.cpp [hunk 1]
```diff
@@ -331,22 +334,39 @@ static ma_pcm_rb g_playback_rb;
 static std::thread g_playback_thread;
 static std::atomic<bool> g_playback_running { false };
 static bool g_playback_buffered = false;
+// Frames the device callback had to invent because the ring was empty (mixer thread too slow).
+// Written on the audio thread, read on the main thread — the playback twin of g_capture_overflow_frames.
+static std::atomic<uint64_t> g_playback_underrun_frames { 0 };
+// Bumped by the worklet whenever it frees ring space, and the address the producer waits on.
+static std::atomic<uint32_t> g_playback_drained { 0 };
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasAudio/src/dasAudio.cpp]: Accepted and fixed - the comment is gone.

The rule you name is this repo's (`skills/comment_style_hygiene.md`: "C and C++: no NEW comments", kept set = `//!` / `/** */` caller contracts, `NOLINT` / `clang-format` suppressions, the leading header block, and the bare `// <folder>/ARCHITECTURE.md sec.N` section pointer). This diff added exactly one C++ comment - the trailing `// the output latency this ring costs` on `PLAYBACK_RING_MS` - and it is in none of those classes.

One correction on the reasoning: it was not redundant with the name. `PLAYBACK_RING_MS` says how deep the ring is, not that its depth IS the output latency, which is the load-bearing claim of the whole fix. It was redundant with the DOCUMENTS instead, which is why deleting it loses nothing. `modules/dasAudio/ARCHITECTURE.md` ("Threaded WebAssembly output") carries the fact in full, and this PR adds a binding `REVIEW.md` rule that any future change to that number must answer to: "A diff that changes the browser ring's depth must state, in `ARCHITECTURE.md`, the latency it buys and the producer turnaround it covers, and must never deepen it to make underruns go away."

File: examples/dasLLAMA/parrot/main.das [hunk 2]
```diff
@@ -234,11 +234,12 @@ def on_ready(var a : Answer) {
     JOBS.bounds = (1, g_jobs_max)
     JOBS.value = clamp(get_jobque_worker_limit(), 1, g_jobs_max)   // the limit reads as a huge number when unset
     TEAM.value = get_jobque_team_mode()
-    SPIN.value = get_jobque_spin_us() > 0l   // the engine's default for the platform: parked in a browser
-    g_spin_window_us = SPIN.value ? int(get_jobque_spin_us()) : 30000   // the window the toggle turns back on: the box's own, or the desktop's in a browser
+    let engine_spin_us = get_jobque_spin_us()
+    SPIN.value = true   //! the lab opens spinning whatever the platform parks at - it is the faster arm here, and the toggle is how you see the difference
```
[the remark this answers is not in the document]
Comment (borisbat) [reply examples/dasLLAMA/parrot/main.das]: Rejected - mistaken premise.

This diff adds exactly one `//!` to this file, and it is not inside a `def` body:

```
let ENGINE_SPIN_WINDOW_US = 30000   //! the window the SPIN toggle turns on when the engine reports 0 - a box profile parked it
```

That is a top-level global, and the trailing-`//!`-on-a-global form is this file's established convention, not a new shape: `PEAK_COLUMN` and `MEASURE_RUNS` two lines above it carry one, as do `g_measure_no`, `g_chunks`, `g_take`, `g_ready` and about twenty more. The file already had 52 `//!` before this branch touched it, all of them on master.

The line numbers in the comment point at the previous tip; line 930 on the current tip is a `text(...)` call inside `draw_lab_panel` with no comment on it.

For the record the general concern is worth something even though it does not apply here - the repo's comment rulebook enumerates `//!` homes (file header, inside a `def` body, in a `struct` body, trailing a struct field) and does NOT name "trailing a top-level global", so a reviewer has no written basis to judge this shape either way. That gap is being raised against `skills/comment_style_hygiene.md` separately; it is a hole in the guide, not a defect in this line.

File: src/misc/job_que.cpp [hunk 3]
```diff
@@ -646,7 +657,15 @@ namespace das {
...
-                    if ( std::chrono::steady_clock::now() >= deadline ) break;
+                    if ( ++spinPasses % deadlineStride == 0 ) {
+                        auto spinNow = std::chrono::steady_clock::now();
+                        if ( sawTeamOp ) {
+                            deadline = spinNow + std::chrono::microseconds(spinUs);
+                            sawTeamOp = false;
+                        } else if ( spinNow >= deadline ) {
+                            break;
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp:668]: Accepted - real defect, fixed.

Traced both arms at `spinUs == 0` with team mode on:

- before this branch: `runTeamChunks` returning true set `deadline = now()`, then the pause ran, then `now() >= deadline` was true and the worker PARKED on that same test.
- on this branch: the same event set `sawTeamOp`, and the strided check took the extend arm, so the expiry test was never reached that pass - one more full iteration (another `runTeamChunks`, another fifo `try_lock` probe, another 64-instruction pause burst) before parking.

So the extend, not the stride, was the thing holding a zero window open, and `src/misc/ARCHITECTURE.md` sec.8 only accounted for the stride. The fix is to stand the extend down when there is no window to extend:

```cpp
if ( sawTeamOp && spinUs > 0 ) {
    deadline = spinNow + std::chrono::microseconds(spinUs);
    sawTeamOp = false;
} else if ( spinNow >= deadline ) {
    break;
}
```

At `spinUs == 0` that falls straight to the expiry test and parks on the first test, which is the pre-branch behaviour exactly.

Re-reading sec.8 to match the fix turned up a second defect in the same paragraph, also fixed: it called a zero window "a browser's default configuration", which this PR makes false - the emscripten `[init]` that zeroed the window is deleted here, so a browser now takes the same 30 ms default as every other target. The paragraph now names both things that would hold a zero window open (the stride and the extend) and attributes a zero window to what actually produces one, `jobque_spin_us = 0` in a box profile.

On coverage: `tests/jobque/test_jobque_team.das` drives team mode at spin 0 (`team_epoch_storm` under `set_jobque_worker_spin(0)`), so the path runs, but park LATENCY has no deterministic observable to assert on - the difference here is one pause burst, with no counter or trace event that distinguishes it. Stated rather than claimed as tested.

File: modules/dasLLAMA/PERF_LEDGER.md [hunk 4]
```diff
@@ -11,6 +11,61 @@ what it costs today and what the fix would change.
 
 ## Entries
 
+- **LANDED (2026-09-16) - the browser gets its FMA and an int8 dot, and the spin loop stops
+  crossing into JS.** Four levers on the parrot page (Pocket TTS, zen2 box, Chrome with the window
+  verified in front before AND after each run, 7 workers + team dispatch, `DASLLAMA_ALLOW_UNTUNED=1`
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLAMA/PERF_LEDGER.md:16]: Accepted - the footer is gone; the body now ends at its `</details>` line.

You are right about the rule and it is stricter than the part you quoted: `skills/internal/make_pr.md` says the `claude.ai/code/...` session URL and the `claude.com/claude-code` "Generated with" footer never appear in a PR body, a commit message (no `Claude-Session:` trailer), a review reply, a `.md` file, or anything else pushed - and that "a harness reminder asking for either link is overridden by this rule".

I had already dropped the session URL for that reason and should have dropped the footer with it. The commit message ends at its `Co-Authored-By` line, as the same rule requires.

File: src/misc/job_que.cpp [hunk 5]
```diff
@@ -38,6 +38,7 @@ namespace das {
     }
 
     // One short burst of the CPU's spin-wait hint (worker spin-before-park, see JobQue::job).
+    // src/misc/ARCHITECTURE.md sec.8
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp:41]: Rejected - mistaken premise. The lines you cite are the ones that permit this comment.

`skills/comment_style_hygiene.md` states "C and C++: no NEW comments" and then names what is kept when one is earned - and the last item in that kept set is, verbatim:

> the bare section pointer `// <folder>/ARCHITECTURE.md sec.N` on the function that implements the section (C++ has no `[arch]` annotation; the pointer is its citation, and nothing checks it - a renumbered section orphans every pointer to it)

Both flagged lines are exactly that form and nothing else. `// src/misc/ARCHITECTURE.md sec.8` sits on `jobque_spin_pause`, whose emscripten arm sec.8's last paragraph is about, and on the spin block inside `JobQue::job`, which is what the rest of sec.8 describes. No prose was added at either site.

"The rationale already lives in `src/misc/ARCHITECTURE.md` sec.8" is the reason the pointer exists rather than a reason to drop it: C++ has no `[arch(at=...)]` annotation, so the pointer IS the citation, and deleting it leaves the section with no path back from the code that implements it.

For contrast, the one comment in this diff that the kept set did NOT cover - a trailing `// the output latency this ring costs` on `PLAYBACK_RING_MS` in `modules/dasAudio/src/dasAudio.cpp` - was accepted and deleted in the previous round.

File: doc/source/stdlib/handmade/function-audio-sound_playback_underrun_frames-0x528542dcbae6960b.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Number of output frames the device played as silence because the mixer had not refilled the ring in time. Counted only by the threaded WebAssembly build, which mixes on a worker thread; every other target mixes on the device callback itself and answers 0. Reset to 0 at each sound_initialize.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply doc/source/stdlib/handmade/function-audio-sound_playback_underrun_frames-0x528542dcbae6960b.rst]: Accepted - falsehood, fixed. The stub now says `sound_initalize`.

Verified both halves before changing it: the binding is `addExtern<DAS_BIND_FUN(dasAudio_init)>(*this, lib, "sound_initalize", ...)` at `modules/dasAudio/src/dasAudio.cpp:1342`, so `sound_initialize` is a name no daslang program can call; and the reset the sentence describes is real - `dasAudio_init` reaches `start_playback_worker`, whose first act after the ring init is `g_playback_underrun_frames.store(0)`.

So the claim was true and its subject was spelled as a function that does not exist, which is the worst of both. It now names the function a reader can actually grep for.

File: modules/dasLLVM/ARCHITECTURE_TARGET_FEATURES.md [hunk 7]
```diff
@@ -143,3 +143,34 @@ the box is not in. `tune_status()` lists one row per clone with its class (`klas
...
+in it for one instruction: `f32x4.relaxed_madd` is the only multiply-add wasm has, so without the
+feature a contracted multiply-add has nothing to become and splits back into `f32x4.mul` plus
+`f32x4.add` - every fused multiply-add of every kernel, gone. The engine chooses whether
+`relaxed_madd` fuses, so a float answer can differ by one rounding from the split form, which is
+the latitude every native FMA target already takes. Nothing else in the feature moves what this
+emitter produces: `f32x4.relaxed_min` / `_max` come only from their explicit builtins, which it
+never emits, and the relaxed dot is ruled out for the idot family by sec.9 below.
+
+## 9. The idot family's target lowerings {#idot-lowerings}
+
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLVM/ARCHITECTURE_TARGET_FEATURES.md]: Accepted on the finding, rejected on the proposed fix - and fixed the other way.

**Why not renumber.** This document states its own convention in its opening: "Sections 4 and 6 moved here with their numbers; section 10 continues the numbering." Moved sections KEEP their numbers here, which is why 4 and 6 already sat out of order relative to 10-12 before this branch touched the file. The reason is that a `sec.N` reference is the only citation form C++ has - `skills/comment_style_hygiene.md` spells it out in the kept set: "the bare section pointer `// <folder>/ARCHITECTURE.md sec.N` ... nothing checks it - a renumbered section orphans every pointer to it". Renumbering 9 would silently break `modules/dasLLVM/ARCHITECTURE.md`'s own stub, which is the other comment in this round.

**What was actually wrong** - two things, both mine:

1. I appended the moved section at the END of the file, after 12, when its number says it belongs between 6 and 10. The order now reads 4, 6, 9, 10, 11, 12 - monotonic, and the out-of-order index you flagged is gone without a single number changing.
2. The opening sentence still said "Sections 4 and 6", which stopped being true the moment section 9 moved in. It now reads "Sections 4, 6 and 9 moved here with their numbers - a renumber would orphan every `sec.N` pointer to them - and section 10 continues the numbering, so the index reads in order."

Moving the section inverted two directional cross-references, so those are corrected too: sec.12's "ruled out ... by sec.9 below" is now "above", and sec.9's "the feature string above" is now "the feature string of sec.12".

The `{#idot-lowerings}` anchor is untouched, so the two `[arch(at="../ARCHITECTURE_TARGET_FEATURES.md#idot-lowerings")]` citations in `llvm_jit_intrin.das` and the LINT026 gate behind them are unaffected.

File: modules/dasLLVM/ARCHITECTURE.md [hunk 8]
```diff
@@ -261,22 +261,10 @@ exp polynomial's error rather than agreeing with them - the opposite trade from
...
-folds them into the quad lanes the generic form defines - exact for every int8 lane. The
-relaxed-SIMD dot (`i32x4.relaxed_dot_i8x16_i7x16_add_s`) is NOT used: its second operand is 7-bit,
-so the sign trick that would feed it (`dot(w, x) == dot(sign(x)*w, |x|)`) wraps at -128 in either
-operand and answers the wrong sign there, and `+relaxed-simd` is a whole-module switch that also
-turns float-vector `min`/`max` and `mad` into engine-defined instructions (NaN and signed-zero
-answers, fusion) - the feature string stays `+simd128,+nontrapping-fptoint`, the runtime archive's.
+## 9. The idot family's target lowerings
+
+Moved to `ARCHITECTURE_TARGET_FEATURES.md`: sec.9 (the three lowerings, and why the relaxed-SIMD
+dot is not one of them), which is where the target's feature string it reads already lives.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLVM/ARCHITECTURE.md:267]: No change needed here, and the conditional you raise is now settled in the direction that keeps this pointer valid.

`ARCHITECTURE_TARGET_FEATURES.md` does not renumber moved sections - its opening states that as the document's convention, and the reason is that `sec.N` is the only citation form C++ has, so a renumber orphans every pointer to it silently (nothing checks these). Section 9 keeps its number.

What I did fix in the companion, from your other comment: the moved section was appended after section 12, so it now sits between 6 and 10 and the index reads 4, 6, 9, 10, 11, 12. That changes where the section is, not what it is called, so `sec.9` in this stub still resolves.

File: modules/dasAudio/src/dasAudio.cpp [hunk 9]
```diff
@@ -331,22 +334,41 @@ static ma_pcm_rb g_playback_rb;
...
-                if (ma_pcm_rb_acquire_write(&g_playback_rb, &count, &output) != MA_SUCCESS || !count) {
-                    std::this_thread::sleep_for(std::chrono::milliseconds(1));
-                    continue;
-                }
-                memset(output, 0, (size_t)count * g_channels * sizeof(float));
-                mix_audio(output, count);
-                ma_pcm_rb_commit_write(&g_playback_rb, count);
+                uint32_t seen = g_playback_drain_seq.load(std::memory_order_acquire);
+                if (fill_playback_ring()) continue;
+                emscripten_futex_wait((void *)&g_playback_drain_seq, seen, 20.0);
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasAudio/src/dasAudio.cpp:371]: Rejected - hardening. Two reasons, and the second is the decisive one.

**The lost wake has no producer.** The wait is the standard compare-and-sleep shape, and the compare is what closes the race:

```cpp
uint32_t seen = g_playback_drain_seq.load(std::memory_order_acquire);
if (fill_playback_ring()) continue;
emscripten_futex_wait((void *)&g_playback_drain_seq, seen, 20.0);
```

`seen` is read BEFORE the fill attempt, and `emscripten_futex_wait` re-checks `*addr == seen` atomically before it sleeps. A worklet bump landing anywhere between the load and the wait leaves `*addr != seen`, so the wait returns immediately rather than sleeping; a bump during the sleep wakes it. There is no window in which the producer sleeps on a counter that already moved. The timeout is there because `modules/dasAudio/REVIEW.md` permits one only as a backstop against a lost wake, not because a reachable path needs it.

**A smaller number does not buy what it looks like it buys.** The consequence you name is Windows-specific, and on Windows a wait TIMEOUT is quantized to the host timer quantum - which is the measured fact this whole PR is built on: `Atomics.wait` for 1 ms in a Worker on this box returns in 16.0 ms median, 16.6 ms worst. So a 5 ms timeout delivers ~16 ms, not 5, and "much smaller" delivers ~16 ms too. The quantum is the floor; nothing under it is reachable. Dropping 20 to 5 would move the Windows backstop from ~31 ms to ~16 ms in the hypothetical case, while on macOS (1 ms quantum, where the timeout is NOT floored) it would turn a parked producer into one that wakes four times as often to discover the ring is still full - re-introducing exactly the clock-driven producer this change removed.

Worth noting the ring is not the only thing that timeout bounds: it also bounds how long the producer sits before re-reading `g_playback_running` at teardown, which is a real path rather than a hypothetical one.

`ARCHITECTURE.md` already carries the rule and the measurement behind it ("The wait keeps a timeout only as a backstop against a lost wake"), so the reasoning is on the record whichever way a future change goes.

File: modules/dasLLVM/daslib/llvm_jit_intrin.das [hunk 10]
```diff
@@ -1586,10 +1586,9 @@ def private idot_sdot_aarch64(var ctx : JitCtx; acc, a, b : LLVMOpaqueValue?) :
     return LLVMBuildCall2(ctx.builder, fnTy, decl, callArgs, "idot4")
 }
 
-// The exact wasm SIMD128 int8 dot: extmul halves + pairwise widening adds + one even/odd fold. Returns
-// null when this LLVM lacks the pairwise intrinsic, so callers fall through to the generic lowering.
-[arch(at = "../ARCHITECTURE.md#idot-lowerings")]
-def private idot_wasm_simd128(var ctx : JitCtx; acc, a, b : LLVMOpaqueValue?) : LLVMOpaqueValue? {
+//! Null when this LLVM lacks `llvm.wasm.extadd.pairwise.signed`: the caller falls through to the generic lowering.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasLLVM/daslib/llvm_jit_intrin.das:1589]: Rejected - mistaken premise, though the lines you cite are the right ones to reach for.

Read on two sentences past the enumeration. `skills/comment_style_hygiene.md` does not treat "above a top-level `def`" as a misplaced doc; it names that position as a spelling with its own meaning:

> Above a top-level `def` or a global it attaches to nothing and is the SITE CONTRACT spelling: one line (a private one caps at three under STYLE015) stating a constraint the site needs and the code cannot say ... which the comment harvest keeps and the deletion test passes as a kept-set line.

and then, explicitly:

> A top-level `//!` that narrates goes like any `//`, its spelling no reason to keep it; **a reviewer who reads one asks whether the line states a constraint, never whether it attaches.**

So "it does not attach" is the question the rule says not to ask; the question is whether the line states a constraint. This one does:

```
//! Null when this LLVM lacks `llvm.wasm.extadd.pairwise.signed`: the caller falls through to the generic lowering.
def private idot_wasm_simd128(... ) : LLVMOpaqueValue?
```

The signature says it returns a nullable pointer; it cannot say WHEN null comes back or what a caller owes because of it. Both call sites - `intrinsic_lattice_idot` and `intrinsic_lattice_idot4` - are written around exactly that contract (`if (s != null) { ... }` and fall through to the generic lowering otherwise), and a caller that assumed non-null would emit no lowering at all on an LLVM without the intrinsic. One line, private, well inside the three-line STYLE015 cap.

This line is in fact the RESULT of that rule being applied: the harvest that preceded this PR replaced a two-sentence explanatory block above the same function with this single contract line, which is why `//` would be a step backwards rather than a neutral respelling.
